### ISSUE TITLE
Comparison in Blue Grey Template is Not Working

### ISSUE THREAD

**Issue by k10blogger** [websvnphp/websvn#112]
Have a look at the snapshot from Elegant Template:
![image](https://user-images.githubusercontent.com/4973182/89707284-569c0080-d98a-11ea-9c3f-af434935dfa8.png)

Same from the Blue Grey Template shows the following message:
![image](https://user-images.githubusercontent.com/4973182/89707298-674c7680-d98a-11ea-8246-e055c0a5cb00.png)

I am not able to figure out why the same comparison works in one template and doesnt in the another.

**Comment by michael-o** [comment]
Did you figure out whether both operations issue the same Subversion command?

**Comment by k10blogger** [comment]
Yeah as far as i can see both operations are issuing the same subversion command as both templates are being called from the same comp.php file. 
The same is visible in calm template as well but not in bluegrey.

Here is the command in elegant and calm
```
svn --non-interactive --config-dir /tmp/websvn --username "k10blogger" --password "xxx" diff "https://svn.code.sf.net/xxx/tags/beta-0.2/@HEAD" "https://svn.code.sf.net/xxx/tags/beta-0.3/@HEAD"
```
Here is the command in the bluegrey
```
svn --non-interactive --config-dir /tmp/websvn --username "k10blogger" --password "xxx" diff "https://svn.code.sf.net/xxx/tags/beta-0.2/@HEAD" "https://svn.code.sf.net/xxx/tags/beta-0.3/@HEAD"
```

They are exactly the same.

**Comment by michael-o** [comment]
If it a ready says "problem with comparison" you need to find the if clause for this. Very strange.

**Comment by ams-tschoening** [comment]
Please describe exactly what you are doing to trigger the problem. Are you starting to choose comparison in `Elegant` and are afterwards just changing the template? Or are you choosing comparison in both templates the same way? Please provide the URLs in both cases as well.

There might simply be something POSTed in one case and not being available if you only switch templates without choosing anything else before.

**Comment by k10blogger** [comment]
It happens in normal operation of bluegrey
In bluegrey view if I check on anything to be compared like shown in the image below:
![image](https://user-images.githubusercontent.com/4973182/89710102-de3f3a80-d99d-11ea-9779-b5e4201f4411.png)
Then hit compare paths i get the error directly.
![image](https://user-images.githubusercontent.com/4973182/89710129-07f86180-d99e-11ea-8851-515599522168.png)

**Comment by ams-tschoening** [comment]
The same happens for me as well and you don't see the warning in some templates simply because there's no corresponding code in those. :-)

    [websvn-test:warning]
      <div id="warning">[websvn:warning]</div>
    [websvn-else]

https://github.com/websvnphp/websvn/blob/master/templates/BlueGrey/compare.tmpl#L20

The problem is with requesting log entries, empty results are returned in my case:

	C:\Users\tschoening>"C:/[...]\svn" --non-interactive --config-dir /tmp/websvn log --xml  --limit 1 -r HEAD:HEAD "file:///C:/[...]/test.txt.pl@HEAD"
	<?xml version="1.0" encoding="UTF-8"?>
	<log>
	</log>

vs.

	C:\Users\tschoening>"C:/[...]\svn" --non-interactive --config-dir /tmp/websvn log --xml  --limit 1 "file:///C:/[...]/test.txt.pl@HEAD"
	<?xml version="1.0" encoding="UTF-8"?>
	<log>
	<logentry
	   revision="261">
	<author>tschoening@am-soft.de</author>
	<date>2012-11-28T17:30:09.280898Z</date>
	<msg>testweise Umbenennung</msg>
	</logentry>
	</log>

The problem is `-r HEAD:HEAD` and should most likely be `-r HEAD:0` instead. This is related to #102.

> It's not a keyword, but I think it's what you want:
> «svn log -r HEAD:0 --limit=1 -- ${TARGET}@HEAD»

https://mail-archives.apache.org/mod_mbox/subversion-users/202005.mbox/%3C20200524220314.1a68eb8a@tarpaulin.shahaf.local2%3E

**Comment by k10blogger** [comment]
@ams-tschoening any solution you might suggest?

**Comment by ams-tschoening** [comment]
1. Change how the warning is handled in the template, so that it's printed ADDITIONALLY as necessary instead of exclusively. The other templates don't even print the warning at all.

2. Make all templates behave the same way, printing the warning ADDITIONALLY if available, but always the other available content as well. Like seems to be the case for most templates already.

3. Recognize the `HEAD:HEAD`-thing and change that to `HEAD:0`, like has been suggested on the SVN mailing list. This should get rid of the warning in the end.

**Comment by k10blogger** [comment]
I looked into the 3rd point. Cause i didnt want to change the html template.
When getting history1 and history2 in the comp.php
We pass text box rev1 for history1 and text box rev2 for history2.

Instead of `$rev1:$rev1` for history1 i did `$rev1:0` for history1 and `$rev2:0` for history2 which seemed not to generate the warning in BlueGrey Template and work as expected in other templates. 
![image](https://user-images.githubusercontent.com/4973182/90181201-b4d34400-ddcd-11ea-9142-4cf574a6d20d.png)

I dont see warning in other templates probably we should add the warning in other template as well.

**Comment by ams-tschoening** [comment]
> Instead of $rev1:$rev1 for history1 i did $rev1:0 for history1 and $rev2:0 for history2[...]

Have a look at `checkRevision` in `comp.php` and test with other keywords than `HEAD` as well. Concrete revision numbers need to still work as well.

> I dont see warning in other templates probably we should add the warning in other template as well.

That's why I added 3 points which need to be done, it's not either of those, but all of them. Normalizing the templates, as covered by the first two points, totally makes sense.

**Comment by k10blogger** [comment]
There is a problem with the keyword PREV,, and Committed.
It seems for some reason the $rev1 and $rev2 are integers only. 
Here is the code snapshot from the comp.php. So no matter what keyword is typed , as it is string it is converted to 0.
![image](https://user-images.githubusercontent.com/4973182/90286010-91be9800-de92-11ea-8a1d-e9704ea595cf.png)
Here if we type HEAD in $rev1 and PREV in $rev2 it is taking zero as integer conversion is done here. 
And as it is taken as 0 the check_revision function changes it to HEAD automatically.

I have made a PR with the two set of changes. That is working with fixed revision numbers and head.

**Comment by ams-tschoening** [comment]
Using the cast to `int` there seems like an error to me. What happens for you if you simply remove that? In the end, the following two lines are called always, normalizing things always:

	$rev1 = checkRevision($rev1);
	$rev2 = checkRevision($rev2);

**Comment by k10blogger** [comment]
I removed the type casting to integer. 
But then again the keyword "PREV" was converted to "HEAD" here in this section:
![image](https://user-images.githubusercontent.com/4973182/90340149-a8d8c380-e013-11ea-89d2-04c7ae904040.png)

I put variable dump before and after this. 
and comparison was with `$rev1 = 'HEAD'` and `$rev2 = 'PREV'`
The $rev2 was converted to HEAD at this point.

**Comment by michael-o** [comment]
@k10blogger Change it to null check, which is a natural value after you have removed the cast to int.

**Comment by k10blogger** [comment]
I did the null check and i think i figured out why it was type cast to integers.
Have a look at this it seems the keyword PREV and COMMITTED are applicable to working copy only and not on the URLS.
![image](https://user-images.githubusercontent.com/4973182/90340718-d45dad00-e017-11ea-914d-b43f963bb7a3.png)

**Comment by michael-o** [comment]
There are now two values. Catch invalid values or leave the behavior undefined.

**Comment by k10blogger** [comment]
@michael-o by catching invalid values do you mean catching and cleaning up the `PREV` and `COMMITTED` keywords?

**Comment by michael-o** [comment]
No only, everything which is not HEAD or 0 to infinity is invalid and could be rejected by some means.

**Comment by k10blogger** [comment]
The 'checkrevision' function does the same. 
![image](https://user-images.githubusercontent.com/4973182/90341656-e4c55600-e01e-11ea-947a-2eccc35ab187.png)

If it is a number it checks for numeric and returns the number.

If it is a string checks for one of the keywords and returns the string to the revision. In string other than keywords returns the HEAD.

**Comment by michael-o** [comment]
There you have it.

**Comment by k10blogger** [comment]
So keeping it unchanged or changing it to following should work:
```
function checkRevision($rev) {
	if (is_numeric($rev) && ((int)$rev > 0)) {
		return $rev;
	}
        return 'HEAD';
}
```

**Comment by michael-o** [comment]
The question is what will Subversion return on invalid values. If it does sanely behave, use the given function, if not use the reduced one.

**Comment by ams-tschoening** [comment]
We shouldn't depend on how SVN behaves, the more interesting question is if `PREV` etc. make sense for us or not. It might simply be that those have been introduced because at some point it wasn't clear that those don't work on URLs or maybe at some point working copies were available. If that has changed right now, there's no need to keep those keywords.

Support for those keywords seems to have been introduced using the following commit and at that point a working copy seems to have been used. `file:///` only make sense with a working copy and that supports `PREV` etc.

    $rawcmd = $config->svn." diff -r$rev1:$rev2 ".quote("file:///".$path1)." ".quote("file:///".$path2);
    $cmd = quoteCommand($rawcmd, true);
    if ($debug) echo "$cmd\n";

https://github.com/websvnphp/websvn/commit/1ee1816b3067982943952198a567206d08480512#diff-adffdff5e4bd35b517b07b0bf87f7637R119
https://github.com/websvnphp/websvn/commit/1ee1816b3067982943952198a567206d08480512

@k10blogger I suggest you check if there's still a working copy involved in case of `comp.php`. If not, we should simply prefer your reduced function. Providing some other keyword than `HEAD` might even be a reason to issue a warning.

**Comment by k10blogger** [comment]
No in current code i dont think any working copy is involved. 
The `file:///` in current code is for the repository path on windows that has a structure something as follows:
```
repos-path
--- conf
--- db
--- hooks
--- locls
--- format
--- readme
```
The PREV keyword will not work with this repository path as well. 
In current code base of `comp.php` there is no reference to file based working copy paths.

**Comment by michael-o** [comment]
As far as I remember, WebSVN never operates on working copies but on repos only -- regardless where they are located.

**Comment by ams-tschoening** [comment]
Then go ahead and remove support for `PREV` like suggested.

**Comment by ams-tschoening** [comment]
Sorry, closed this by accident and merged PR #113 already because I forgot about removing `PREV`. I suggest creating a new PR simply.
